### PR TITLE
Add trace WorkflowExecutionUpdate and its worker

### PR DESCRIPTION
**Pull request by Sikian**
## What was changed
Add `GetWorkflowExecutionUpdates` function that uses `GetWorkflowHistory` to aggregate the workflow execution state. New updates can be retrieved using the HasNext()/Next() pattern, I hope I implemented this correctly.

This uses a workerpool to spawn new aggregation jobs for child workflows. I decided to use `pond` due to being able to easily wait for all jobs to complete (instead of closing and waiting as other workerpools do) and to easily propagate contexts and errors. Let me know if there's any problem with this.

This also includes adding a mutex lock to the `WorkflowExecutionState` maps storing the child workflows/activity/timers to avoid concurrent read/writes.

## Why?
This is the main engine of the workflow tracing. 

Next step will be the frontend, which uses `GetWorkflowExecutionUpdates` to get the latest `WorkflowExecutionState` and display it.

## Checklist

There are tests for `GetWorkflowExecutionUpdates` which cover the whole functionality. Let me know if you'd like more specific tests for the different components.

### CODE REVIEW

File: trace/workflow_execution_update.go
Comment (bergundy) [inline trace/workflow_execution_update.go:22]: Is the separation of interface and implementation for mocking purposes?

File: trace/workflow_execution_update_test.go
Comment (bergundy) [inline trace/workflow_execution_update_test.go:35]: I'd prefer integration tests over unit tests with mocks to get better confidence that this code works with a real server.

File: trace/execution_state.go
Comment (bergundy) [reply trace/execution_state.go:78]: To be safe, I think we should lock the entire struct on any field access.
I haven't looked super closely at the access patterns but since this is all meant to be accessed concurrently, I'd rather be on the safe side.

**Review by bergundy — COMMENTED** [review]
Thanks!
Overall LGTM.

File: trace/workflow_execution_update_test.go
Comment (Sikian) [reply trace/workflow_execution_update_test.go:35]: Good idea. Let me get to it 👍

File: trace/workflow_execution_update.go
Comment (Sikian) [reply trace/workflow_execution_update.go:22]: Not really. I was just following the iterator pattern but didn't have a better name for the implementation. Happy to change this if it's not the right approach.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:116]: Probably not, I'll remove it.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:78]: That makes sense, thanks for the suggestion.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:82]: Renamed this to `childWorkflowMap`.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:152]: Wasn't aware of this, thanks!

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:158]: The activity map is accessed in the following line.

File: trace/workflow_execution_update.go
Comment (Sikian) [reply trace/workflow_execution_update.go:6]: I'm not sure if there's a vanilla workerpool. I'm happy to implement it myself (I did this initially) but I believe it might be better in the long run to use a well-tested workerpool.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:217]: Good point.

File: trace/workflow_execution_update.go
Comment (Sikian) [reply trace/workflow_execution_update.go:104]: I think HasNext() doesn't return any error for `HistoryEventIterator`. However, I can see that that implementation will advance the index every time `Next()` is called rather than in `HasNext()`. I can change that if that makes more sense.

File: trace/workflow_execution_update_test.go
Comment (Sikian) [reply trace/workflow_execution_update_test.go:35]: I've been giving this some thought and I'm not sure I understand what you're trying to convey. Is your suggestion to setup a server with fixture responses and test against that?

If that is the case: while I agree that a server would make these tests closer to real, I think leveraging the sdk mocks can make the tests easier to maintain (no server to maintain) and we can get exactly the response we want for each test. Additionally, failures in `DescribeWorkflowExecution` and `GetWorkflowHistory` in the sdk would bleed into the tests here.

In any case, I'm happy to review this and add more tests to better cover this functionality.

File: trace/execution_state.go
Comment (bergundy) [inline trace/execution_state.go:78]: Are these the only properties that will be accessed concurrently?
If you're certain of this, let's keep the way it is but otherwise, I'd put a lock on the entire struct.

File: trace/execution_state.go
Comment (bergundy) [inline trace/execution_state.go:142]: Don't we need to lock the map here?

File: trace/execution_state.go
Comment (bergundy) [inline trace/execution_state.go:174]: You should move the lock to the top of this function.

File: trace/execution_state.go
Comment (bergundy) [inline trace/execution_state.go:194]: Don't need all of this anymore.

File: trace/execution_state.go
Comment (bergundy) [inline trace/execution_state.go:193]: Leftover?

File: trace/execution_state.go
Comment (bergundy) [reply trace/execution_state.go:318]: This comment applies to anywhere you access the maps in this function.

File: trace/workflow_execution_update.go
Comment (bergundy) [reply trace/workflow_execution_update.go:6]: I was thinking just use goroutines and a semaphore.

File: trace/workflow_execution_update.go
Comment (bergundy) [reply trace/workflow_execution_update.go:22]: Let's merge the interface and implementation.

File: trace/workflow_execution_update.go
Comment (bergundy) [reply trace/workflow_execution_update.go:104]: Yeah, you could panic to be on the safe side but it's fine for internal use.

File: trace/workflow_execution_update_test.go
Comment (bergundy) [reply trace/workflow_execution_update_test.go:35]: My suggestion is that we test real server responses, that would mean starting a workflow and getting it to a specific state.

I think these unit tests are fine but I'd want at least a single integration test, can wait for when we implement the display logic.

File: trace/workflow_state_worker.go
Comment (bergundy) [inline trace/workflow_state_worker.go:38]: I'd rather just return the error here instead of silently ignoring it. There's no good reason for this call to fail and if it does it'll be hard to debug.

File: trace/workflow_state_worker.go
Comment (bergundy) [inline trace/workflow_state_worker.go:114]: Looks like `wf` is accessed concurrently, so I'm now convinced we should have a lock for the entire struct.

Comment (Sikian) [reply trace/workflow_state_worker.go:114]: That's what the lock in `GetChildWorkflowByEventId` is for.

File: trace/workflow_state_worker.go
Comment (Sikian) [reply trace/workflow_state_worker.go:38]: I disagree, but I'll update it as required.

File: trace/workflow_execution_update_test.go
Comment (Sikian) [reply trace/workflow_execution_update_test.go:35]: That's a significant investment into some tests which will probably not provide better coverage. If you believe it's going to add value, I invite you to implement it in the rest of the tool.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:78]: There's no problem with properties being accessed concurrently, only maps.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:142]: Done 👍

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:193]: Removed, thanks.

File: trace/execution_state.go
Comment (Sikian) [reply trace/execution_state.go:142]: Done, thanks 👍

File: trace/workflow_state_worker.go
Comment (Sikian) [reply trace/workflow_state_worker.go:20]: Added

Comment (Sikian) [reply trace/workflow_state_worker.go:20]: Done 👍

File: trace/execution_state.go
Comment (bergundy) [reply trace/execution_state.go:78]: It depends if we care about multiple properties being updated atomically.
I generally would like to make less assumptions and be on the safe side.

File: trace/workflow_execution_update_test.go
Comment (bergundy) [reply trace/workflow_execution_update_test.go:35]: Yes, we need more integration coverage in general but I won't block the PR on this.

File: trace/workflow_state_worker.go
Comment (bergundy) [reply trace/workflow_state_worker.go:118]: Yes but then you pass `wf` to `NewWorkflowStateJob` unlocked.

Again this might not matter today since we update a couple of attributes (`HistoryLength` and `IsArchived`) that may not need to be atomically updated but, as mentioned above, I'd rather be on the safe side and make sure all access to this struct is serialized.

File: trace/workflow_state_worker.go
Comment (Sikian) [reply trace/workflow_state_worker.go:20]: `	foldStatus []enums.WorkflowExecutionStatus`

File: trace/workflow_execution_update.go
Comment (bergundy) [reply trace/workflow_execution_update.go:30]: Not blocking but would be nice to address before merging.

File: trace/workflow_state_worker.go
Comment (bergundy) [reply trace/workflow_state_worker.go:20]: So what's `fetchAll` for?

Comment (bergundy) [reply trace/workflow_state_worker.go:20]: I see, I thought we agreed to remove the `--all` flag in favor of the individual statuses.

File: trace/workflow_state_worker.go
Comment (bergundy) [reply trace/workflow_state_worker.go:118]: Each state struct would have its own lock, you wouldn't hold the same lock for all workflows.
Again, I'd rather have code that makes less assumptions about concurrent usage but I think we can move on to the next PR.

**Review by bergundy — APPROVED** [review]
Left a few more comments.
Can address in the next PR.
Let's merge this and move forward.

File: trace/workflow_state_worker.go
Comment (Sikian) [reply trace/workflow_state_worker.go:20]: We're also using the `--no-fold` to disable the behavior: https://github.com/temporalio/cli/blob/main/workflow/workflow.go#L381-L385

Comment (bergundy) [reply trace/workflow_state_worker.go:20]: Right, I forgot that what we settled on.

**Comment by bergundy** [comment]
@feedmeapples looks like CI is failing on test-install, I'm merging this as the failures aren't related to the change.

**Pull request merged**
